### PR TITLE
Output more information about local numbers on console

### LINKED ISSUE #18492: Output min/max number of elements/nodes in outputMeshInformation instead of the numbers on head processor
## Reason
I met a case where the partitioning is severely unbalanced but did not realize until checking it very carefully. This makes me think that possibly it is better to output the min/max number of elements/nodes in `outputMeshInformation` than to output just the two numbers on the head processor. To make `--keep-cout` (which output the local numbers of all processors) still useful, we can keep the old local numbers and add outputs of the min/max on head processor.

## Design
Change the function `outputMeshInformation` in `ConsoleUtils.C` a little.

## Impact
Should only affect console outputting and only making the console outputting more informative.

### PR DESCRIPTION
**Pull request by YaqiWang**
Close #18492.

This PR also comes with two minor fixes for closing #18495 and #16106.

### CODE REVIEW

**Comment by GiudGiud** [comment]
I think this is a job for postprocessors, eg NumElems, NumDOFs, NumNodes. Very few people need to know the average and the RMS, or the min, number of nodes.

**Comment by GiudGiud** [comment]
Still undecided though. It doesn't look super good on the console. And your goals would be achieved by simply outputting the max below global and local.

**Comment by YaqiWang** [comment]
Yeah, on sample output is like this:
```
Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   
    Total:                 3549
    Local:                 546 (Min: 360, Max: 546, Avg: 443, RMS: 446)
  Elems:                   
    Total:                 2880
    Local:                 360 (Min: 360, Max: 360, Avg: 360, RMS: 360)
  Num Subdomains:          3
  Num Partitions:          8
  Partitioner:             metis
```
We can have something like if you guys think it better:
```
  Nodes:                   
    Total:                 3549
    Local:                 546
    Distribution:      Min: 360, Max: 546, Avg: 443, RMS: 446
```

**Comment by loganharbour** [comment]
Are the other quantities really necessary? My thoughts are to use the following:

`n_processors() > 1`:

```
Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   
    Total:                 3549
    Local Min/Max:         360/546
  Elems:                   
    Total:                 2880
    Local Min/Max:         360/360
  Num Subdomains:          3
  Num Partitions:          8
  Partitioner:             metis
```

`n_processors() == 1`:

```
Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   3549
  Elems:                   2880
  Num Subdomains:          3
```

**Comment by loganharbour** [comment]
I think the thing that I'm set on the most is that there's no need to print the information about processor 0 (local elems/nodes) if we expand this

**Comment by friedmud** [comment]
I like the idea of outputting (a bit) more information here.

I like the look of Logan's better, but we could add average as well if @YaqiWang really wants it.

One suggestion is that it really is useful to see the counts for MPI rank 0 as well.  The reason why is that it can be useful to count to make sure that what you are running is working.  But - I may be the only one interested - so it's probably not necessary.

**Comment by GiudGiud** [comment]
I dont think we need all 4 to raise awareness about imbalance. 
max and average suffice. max and min could work too. rms could sub out for average. Average is kind of very easy to compute already.

I like this format the best:
```
  Nodes:                   
    Total:                 3549
    Local:                360
    Min/Max:            200 / 546

```
or Logan's with Local
```
  Nodes:                   
    Total:                 3549
    Local/Avg/Max:  360 / 400 / 546
```

**Comment by YaqiWang** [comment]
I think outputting `local` numbers could be still useful because we might be outputting related local stuff like petsc vector size, etc on local as well. I like Logan's suggestion for making one processor output cleaner. How about combining all this into:
```
Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   
    Total:                 3549
    Local:                 546
    Min/Max/Avg/RMS:         360/546/443/446
  Elems:                   
    Total:                 2880
    Local:                 360
    Min/Max/Avg/RMS:         360/360/360/360
  Num Subdomains:          3
  Num Partitions:          8
  Partitioner:             metis
```

**Comment by YaqiWang** [comment]
With the latest code, the output is like:
```
Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   
    Total:                 3549
    Local:                 417
    Min/Max/Avg/RMS:       245/423/322/327
  Elems:                   
    Total:                 2880
    Local:                 255
    Min/Max/Avg/RMS:       255/269/261/261
  Num Subdomains:          3
  Num Partitions:          11
  Partitioner:             metis
```
or
```
Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   3549
  Elems:                   2880
  Num Subdomains:          3
```

**Comment by YaqiWang** [comment]
Actually `Avg` can be deduced with the total number divided by the number of processors, so it is a redundant info. I guess we can still keep it for showing how far the min and max deviate from the average by just looking at the line.

**Comment by YaqiWang** [comment]
> I dont think we need all 4 to raise awareness about imbalance.
> max and average suffice. max and min could work too. rms could sub out for average. Average is kind of very easy to compute already.

Actually max and min are equally important I think. We can keep average to see clearly how the imbalance is. RMS is less important. We need a little math to convert it to variance that carries a little information on if the distribution is even between min and max. I am ok to remove RMS if you have strong opinions on it.

**Comment by loganharbour** [comment]
With the conversation above (on avg being deducible and rms not as important), I'm for either `Local Min/Max:` or `Local/Min/Max`. I'd prefer the former, but that one's on @friedmud

**Comment by GiudGiud** [comment]
fyi in a high element scenario it looks like this
```
Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   
    Total:                 3932094820
    Local/Min/Max: 417234430/302394021/204909839
  Elems:                   
    Total:                 288089675
    Local/Min/Max:  345676255/987456098/879098457
  Num Subdomains:          3
  Num Partitions:          11
  Partitioner:             metis

**Comment by loganharbour** [comment]
300m local elements? No way. Regardless, another reason to drop local :)

**Comment by GiudGiud** [comment]
true true
just having min/max doesn't convey the information that we are talking about per-rank quantities though

```
Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   
    Total:                 3932094820
    Local/Min/Max: 417234430/302394021/204909839
  Elems:                   
    Total:                 288089675
    Local/Min/Max:  34567/98745/87909
  Num Subdomains:          3
  Num Partitions:          1100
  Partitioner:             metis

**Comment by loganharbour** [comment]
This summary is supposed to be a summary, keep that in mind. If we want descriptive capability, we should add it as an option.

**Comment by GiudGiud** [comment]
I fully agree with that. We should add as little possible information as possible here, no redundancy especially, AND it should be absolutely clear what the min/max/local etc are referring to.

**Comment by YaqiWang** [comment]
Ok guys. This is the new look:
```
Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   
    Total:                 3549
    Local:                 417
    Local min/max/average: 245/423/322
  Elems:                   
    Total:                 2880
    Local:                 255
    Local min/max/average: 255/269/261
  Num Subdomains:          3
  Num Partitions:          11
  Partitioner:             metis
```
I want to keep average because that can show how far min and max is from it by just looking at this output line.

**Comment by loganharbour** [comment]
How about:
```
Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   
    Total:                 3549
    Local:                 417
    Min/Max/Avg:           245/423/322
  Elems:                   
    Total:                 2880
    Local:                 255
    Min/Max/Avg:           255/269/261
  Num Subdomains:          3
  Num Partitions:          11
  Partitioner:             metis
```

Something about seeing local twice makes it seem repetitive in my eyes.

File: framework/src/outputs/ConsoleUtils.C [hunk 1]
```diff
@@ -81,22 +81,64 @@ outputMeshInformation(FEProblemBase & problem, bool verbose)
...
-      << std::setw(console_field_width) << "    Local:" << mesh.n_local_nodes() << '\n'
-      << std::setw(console_field_width) << "  Elems:" << '\n'
-      << std::setw(console_field_width) << "    Total:" << mesh.n_active_elem() << '\n'
-      << std::setw(console_field_width) << "    Local:" << mesh.n_active_local_elem() << '\n';
+  dof_id_type min_nnodes = std::numeric_limits<dof_id_type>::max(), max_nnodes = 0;
+  dof_id_type min_nelems = std::numeric_limits<dof_id_type>::max(), max_nelems = 0;
+  if (mesh.n_processors() > 1)
+  {
+    std::vector<dof_id_type> n_local_nodes;
+    mesh.comm().gather(0, mesh.n_local_nodes(), n_local_nodes);
```
Comment (fdkong) [inline framework/src/outputs/ConsoleUtils.C:89]: I would just use  `mesh.comm().max()` and `mesh.comm().min()`

**Review by fdkong — CHANGES_REQUESTED** [review]

File: framework/src/outputs/ConsoleUtils.C (hunk 1 above)
Comment (YaqiWang) [reply framework/src/outputs/ConsoleUtils.C:89]: lol. You are right. I was evaluating RMS using this before. But if we just need max and min, we do not need this.

**Comment by YaqiWang** [comment]
In distributed mesh mode, `mesh.n_active_elem()` is collective... After the tests pass, this should be ready for review again.

**Comment by YaqiWang** [comment]
More comments?

**Comment by GiudGiud** [comment]
the people asked for the median. I did some googling there's an algorithm
https://www.quora.com/Distributed-Algorithms/What-is-the-distributed-algorithm-to-determine-the-median-of-arrays-of-integers-located-on-different-computers/answer/Michael-Harris?srid=Oda1&share=1

**Comment by YaqiWang** [comment]
@GiudGiud looks like the poll on avg and median are even. I hate to upset you but I do not have time to add finding median algorithm in moose utilities. I think `avg` is good enough. We can have the postprocessor later to evaluate fancy stuff like RMS, median, etc.

**Comment by fdkong** [comment]
>@GiudGiud looks like the poll on avg and median are even. I hate to upset you but I do not have time to add finding median algorithm in moose utilities. I think avg is good enough. We can have the postprocessor later to evaluate fancy stuff like RMS, median, imbalance, std derivation, etc.

In my humble opinion, we might not report median for this use case. First of all, we get `avg` for free. If we want `median`, we have to do at least one `gather`  that simply overkill. Second, when people report imbalance information in literature, people do `max/avg `. Third, when you look at `median`, you do not really get the information of `imbalance`.

**Review by fdkong — APPROVED** [review]
Looks good to me

**Comment by GiudGiud** [comment]
Looks like the poll changed somehow ^^.
ok fine with average.

**Review by GiudGiud — APPROVED** [review]
Thanks for the contribution!

**Comment by YaqiWang** [comment]
Thanks!

**Pull request merged**